wolfcrypt/src/port/nxp/dcp_port.c

```diff
@@ -0,0 +1,515 @@
+    #include <wolfcrypt/src/misc.c>
+#endif
+
+#include <wolfssl/wolfcrypt/aes.h>
+#include <wolfssl/wolfcrypt/sha.h>
+#include <wolfssl/wolfcrypt/sha256.h>
+#include <wolfssl/wolfcrypt/error-crypt.h>
+
+#ifdef WOLFSSL_IMXRT_DCP
+#if defined(__DCACHE_PRESENT) && (__DCACHE_PRESENT == 1U) && defined(DCP_USE_DCACHE) && (DCP_USE_DCACHE == 1U)
+#error "DCACHE not supported by this driver. Please undefine DCP_USE_DCACHE."
+#endif
+
+#ifndef DCP_USE_OTP_KEY
+#define DCP_USE_OTP_KEY 0 /* Set to 1 to select OTP key for AES encryption/decryption. */
+#endif
+
+#include "fsl_device_registers.h"
+#include "fsl_debug_console.h"
+#include "fsl_dcp.h"
+
+static int dcp_is_initialized = 0;
+
+#ifndef SINGLE_THREADED
+static wolfSSL_Mutex dcp_mutex;
+static void dcp_lock(void)
+{
+    if (!dcp_is_initialized)
+        return;
+    wc_LockMutex(&dcp_mutex);
+}
+
+static void dcp_unlock(void)
+{
+    if (!dcp_is_initialized)
+        return;
+    wc_UnLockMutex(&dcp_mutex);
+}
+#else
+#define dcp_lock() do{}while(0)
+#define dcp_unlock() do{}while(0)
+#endif
+
+#if DCP_USE_OTP_KEY
+typedef enum _dcp_otp_key_select
+{
+    kDCP_OTPMKKeyLow  = 1U, /* Use [127:0] from snvs key as dcp key */
+    kDCP_OTPMKKeyHigh = 2U, /* Use [255:128] from snvs key as dcp key */
+    kDCP_OCOTPKeyLow  = 3U, /* Use [127:0] from ocotp key as dcp key */
+    kDCP_OCOTPKeyHigh = 4U  /* Use [255:128] from ocotp key as dcp key */
+} dcp_otp_key_select;
+#endif
+
+#if DCP_USE_OTP_KEY
+static status_t DCP_OTPKeySelect(dcp_otp_key_select keySelect)
+{
+    status_t retval = kStatus_Success;
+    dcp_lock();
+    if (keySelect == kDCP_OTPMKKeyLow)
+    {
+        IOMUXC_GPR->GPR3 &= ~(1 << IOMUXC_GPR_GPR3_DCP_KEY_SEL_SHIFT);
+        IOMUXC_GPR->GPR10 &= ~(1 << IOMUXC_GPR_GPR10_DCPKEY_OCOTP_OR_KEYMUX_SHIFT);
+    }
+
+    else if (keySelect == kDCP_OTPMKKeyHigh)
+    {
+        IOMUXC_GPR->GPR3 |= (1 << IOMUXC_GPR_GPR3_DCP_KEY_SEL_SHIFT);
+        IOMUXC_GPR->GPR10 &= ~(1 << IOMUXC_GPR_GPR10_DCPKEY_OCOTP_OR_KEYMUX_SHIFT);
+    }
+
+    else if (keySelect == kDCP_OCOTPKeyLow)
+    {
+        IOMUXC_GPR->GPR3 &= ~(1 << IOMUXC_GPR_GPR3_DCP_KEY_SEL_SHIFT);
+        IOMUXC_GPR->GPR10 |= (1 << IOMUXC_GPR_GPR10_DCPKEY_OCOTP_OR_KEYMUX_SHIFT);
+    }
+
+    else if (keySelect == kDCP_OCOTPKeyHigh)
+    {
+        IOMUXC_GPR->GPR3 |= (1 << IOMUXC_GPR_GPR3_DCP_KEY_SEL_SHIFT);
+        IOMUXC_GPR->GPR10 |= (1 << IOMUXC_GPR_GPR10_DCPKEY_OCOTP_OR_KEYMUX_SHIFT);
+    }
+
+    else
+    {
+        retval = kStatus_InvalidArgument;
+    }
+    dcp_unlock();
+    return retval;
+}
+#endif
+
+static const int dcp_channels[4] = {
+    kDCP_Channel0,
+    kDCP_Channel1,
+    kDCP_Channel2,
+    kDCP_Channel3
+};
+static int dcp_status[4] = {0, 0, 0, 0};
+
+static int dcp_get_channel(void)
+{
+#ifdef SINGLE_THREADED
+    return dcp_channels[0];
+#else
+    int i;
+    for (i = 0; i < 4; i++) {
+        if (dcp_status[i] == 0) {
+            dcp_status[i]++;
+            return dcp_channels[i];
+        }
+    }
```

Comment:
mutex protection: OK.
semaphore: I understand that it would be nice in a multi-threaded environment to hold the thread until a channel/slot is available. OTOH, this is out of our scope, and could be implemented in a different layer. In my opinion it is better to explicitly notify the caller when no slots are available via `WC_PENDING_E`.